Add array indexing to NodeCollections

hakonsbm opened the pull request:
This PR adds the ability to index NodeCollections with arrays of integers or bools, similar to [advanced indexing in NumPy](https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#advanced-indexing). The result is a new `NodeCollection` with the selected nodes. Note that a `NodeCollection` can only contain unique node IDs. Array indexing supports lists, tuples, and NumPy arrays as indices.

Example:
```Python
n = nest.Create('iaf_psc_alpha', 5)
n[[1, 2]]  # -> NodeCollection with node IDs 2 and 3
n[(0, 1, 3, 4)]  # -> NodeCollection with node IDs 1, 2, 4, and 5
n[[True, False, True, False, True]]  # -> NodeCollection with node IDs 1, 3, and 5
n[[2, 3, 3, 4]]  # Error: NodeCollection cannot contain non-unique node IDs
```

jougs commented:
@hakonsbm: I would only review once you addressed the comments by @alberto-antonietti and fixed the conflicts, OK?

hakonsbm commented on pynest/pynestkernel.pyx:
```diff
@@ -247,6 +247,42 @@ cdef class NESTEngine(object):
+                new_nc_datum = node_collection_array_index(nc_datum, array_long_ptr, len(array))
+                return sli_datum_to_object(new_nc_datum)
+            else:
+                raise TypeError('array must be a NumPyArray of ints or bools, got {}'.format(array.dtype))
```
Fixed.

hakonsbm commented on pynest/pynestkernel.pyx:
```diff
@@ -247,6 +247,42 @@ cdef class NESTEngine(object):
+
+        if not (isinstance(node_collection, SLIDatum) and (<SLIDatum> node_collection).dtype == SLI_TYPE_NODECOLLECTION.decode()):
+            raise TypeError('node_collection must be a NodeCollection, got {}'.format(type(node_collection)))
+        if not isinstance(array, numpy.ndarray):
```
Good point, I have added it.

hakonsbm commented on nestkernel/nest.cpp:
```diff
@@ -425,4 +425,38 @@ apply( const ParameterDatum& param, const DictionaryDatum& positions )
+  std::vector< index > node_ids;
+  node_ids.reserve( n );
+
+  for ( auto node_ptr = array; node_ptr != array + n; ++node_ptr )
```
The idea is that we don't care about the type of `node_ptr` as long as it's the same as the type of `array`. If the type of `array` is changed, it only has to change in the definition of the function. Ensuring that they always have the same type can avoid some nasty bugs down the line if the type of `array` is changed.

alberto-antonietti replied:
OK, got it.

hakonsbm commented on nestkernel/nest.cpp:
```diff
@@ -425,4 +425,38 @@ apply( const ParameterDatum& param, const DictionaryDatum& positions )
+  node_ids.reserve( n );
+
+  auto nc_it = node_collection->begin();
+  for ( auto node_ptr = array; node_ptr != array + n; ++node_ptr, ++nc_it )
```
Same as above.

hakonsbm commented on nestkernel/node_collection.cpp:
```diff
@@ -158,7 +170,7 @@ NodeCollection::create_( const std::vector< index >& node_ids )
   std::vector< NodeCollectionPrimitive > parts;
 
-  index old_node_id = 0;
+  index old_node_id = current_first;
```
True, it's not relevant for the array indexing. I can take it out and create a separate PR for it.

hakonsbm replied:
The changes are now in #1536.

hakonsbm commented:
@alberto-antonietti Thanks for your review! I have addressed most of your comments, still working on the last one.

> Another comment not strictly related to your changes.
> When working with lists or arrays, if an index outside of the NodeCollection is used, then the right error is thrown:
>
> (...)
>
> I do not know if this PR is the place where to fix this inconsistency. Otherwise, I can create an ad-hoc PR for this.

I agree the error message could be more informative here. If you create a PR for it, I'll be happy to review it.
